consensus/src/threshold_simplex/signing_scheme/ed25519.rs

```diff
@@ -0,0 +1,791 @@
+//! Ed25519 quorum signature implementation of the signing scheme abstraction.
+//!
+//! This module batches ordinary Ed25519 signatures from a quorum of participants
+//! and packages them as certificates that satisfy the generic consensus interface.
+
+use crate::threshold_simplex::{
+    signing_scheme::{finalize_namespace, notarize_namespace, nullify_namespace},
+    types::{Participants, SigningScheme, Vote, VoteContext, VoteVerification},
+};
+use bytes::{Buf, BufMut};
+use commonware_codec::{Encode, EncodeSize, Error, Read, ReadRangeExt, Write};
+use commonware_cryptography::{
+    ed25519::{Batch, PrivateKey, PublicKey, Signature as Ed25519Signature},
+    BatchVerifier, Digest, Signer as _, Verifier as _,
+};
+use rand::{CryptoRng, Rng};
+use std::collections::BTreeSet;
+
+/// Ed25519 implementation of the [`SigningScheme`] trait.
+///
+/// The scheme keeps the participant ordering plus (optionally) the local private
+/// key so it can produce votes as well as batch-verify signatures from peers.
+#[derive(Clone, Debug)]
+pub struct Scheme {
+    /// Participant set (sorted) used for signer indexing and batch verification.
+    participants: Participants<PublicKey>,
+    /// Optional local signing key paired with its participant index.
+    signer: Option<(u32, PrivateKey)>,
+}
+
+impl Scheme {
+    /// Creates a new scheme instance with the provided key material.
+    ///
+    /// * `participants` - ordered validator set used for verification.
+    /// * `private_key` - optional secret key enabling signing capabilities.
+    pub fn new(participants: Vec<PublicKey>, private_key: PrivateKey) -> Self {
+        let participants = Participants::from(participants);
+
+        let signer = participants
+            .signer_index(&private_key.public_key())
+            .map(|index| (index, private_key));
+
+        Self {
+            participants,
+            signer,
+        }
+    }
+
+    /// Builds a pure verifier that can authenticate votes without signing.
+    pub fn verifier(participants: Vec<PublicKey>) -> Self {
+        Self {
+            participants: participants.into(),
+            signer: None,
+        }
+    }
```

Comment:
nit: I think in these function signatures you should technically use `participants: impl Into<Participants<PublicKey>>` to allow more types to be used here (including `Participants<PublicKey>` directly)